### DISCUSSION TITLE
Sharing low level code with x3d.py

### DISCUSSION THREAD

**Discussion by coderextreme** [pygfx/pygfx#551] [Ideas]
I’ve been looking for a python package to provide geometry, appearance, materials, textures, lighting, camera, grouping, transforms—stuff like Three.JS and Babylon.JS provides, but in Python.    I want to use the latest WebGPU and WGPU APIs.

I would like to share a core API including the above.  That is, I will provide the Model in Model-View-Controller, and the core API will provide the rendered View.  Is there a possibility for us to collaborate?  My current model is based on X3D Python Scene Access Interface Library, X3DPSAIL, and is the “x3d” python package installable with pip.  I don’t have a current renderer/View in python.  That’s what i am seeking.

**Comment by almarklein** [comment]
Hi John,

> I’ve been looking for a python package to provide geometry, appearance, materials, textures, lighting, camera, grouping, transforms—stuff like Three.JS and Babylon.JS provides, but in Python. I want to use the latest WebGPU and WGPU APIs.

That sounds like pygfx is a good match.

> I would like to share a core API including the above. That is, I will provide the Model in Model-View-Controller, and the core API will provide the rendered View. Is there a possibility for us to collaborate?

You can use pygfx for your rendering needs. You speak of sharing a core API and collaborating, but I am not sure what it is you're asking of us :) Could you please clarify?

**Reply by coderextreme** [reply]
The core API would do the rendering for the "x3d" python package, which should be installable with pip/pip3?  Maybe it was with withdrawn?  It can be made to work with XML/X3DOM rendering, I think. https://www.web3d.org/x3d/stylesheets/python/python.html

**Reply by almarklein** [reply]
As I understand it, x3d is a Python module to access 3D models in a standardized way. As far as I can tell it does not render anything. If you want you can use pygfx to render objects read via x3d.

**Reply by coderextreme** [reply]
X3d.py does not do rendering at this time, but it provides XML export to things like X3DOM and X_ITE (JavaScript based).  But providing a rendering package written in python would probably be a good fit.
